File: orchestrator-service/src/main/java/com/credit/orchestrator/validator/CedulaValidator.java
package com.credit.orchestrator.validator;

import java.util.List;

/**
 * Validador de cédula ecuatoriana.
 * 
 * SRP: Única responsabilidad - validar cédula.
 * OCP: Extiende mediante el método validateField.
 */
public class CedulaValidator implements FieldValidator<String> {
    
    private static final int MIN_PROVINCE = 1;
    private static final int MAX_PROVINCE = 31;
    private static final int[] COEFFICIENTS = {2, 1, 2, 1, 2, 1, 2, 1, 2};

    @Override
    public String validate(String cedula) {
        if (cedula == null || cedula.length() != 10) {
            return "Cédula ecuatoriana inválida (10 dígitos, módulo 10)";
        }
        if (!cedula.matches("\\d{10}")) {
            return "Cédula ecuatoriana inválida (10 dígitos, módulo 10)";
        }
        int provinceCode = Integer.parseInt(cedula.substring(0, 2));
        if (provinceCode < MIN_PROVINCE || provinceCode > MAX_PROVINCE) {
            return "Cédula ecuatoriana inválida (código de provincia no válido)";
        }
        int thirdDigit = Character.getNumericValue(cedula.charAt(2));
        if (thirdDigit > 5) {
            return "Cédula ecuatoriana inválida (tercer dígito debe ser 0-5)";
        }
        int sum = 0;
        for (int i = 0; i < 9; i++) {
            int digit = Character.getNumericValue(cedula.charAt(i));
            int product = digit * COEFFICIENTS[i];
            if (product >= 10) {
                product -= 9;
            }
            sum += product;
        }
        int verifier = Character.getNumericValue(cedula.charAt(9));
        int expected = (sum % 10 == 0) ? 0 : 10 - (sum % 10);
        if (verifier != expected) {
            return "Cédula ecuatoriana inválida (dígito verificador no coincide)";
        }
        return null;
    }
    
    public void validateField(String value, String fieldName, List<String> errors) {
        String error = validate(value);
        if (error != null) {
            errors.add(error);
        }
    }

    public static boolean isValid(String cedula) {
        return new CedulaValidator().validate(cedula) == null;
    }
}

File: frontend/src/utils/cedulaValidator.ts
const PROVINCE_CODES = new Set([
  "01","02","03","04","05","06","07","08","09","10",
  "11","12","13","14","15","16","17","18","19","20",
  "21","22","23","24","25","26","27","28","29","30",
  "31",
]);

export function validateCedulaEcuadoriana(cedula: string): string | null {
  const clean = cedula.replace(/\D/g, "");

  if (clean.length !== 10) return "La cédula debe tener 10 dígitos";
  if (!/^\d{10}$/.test(clean)) return "La cédula debe contener solo dígitos";

  const province = clean.substring(0, 2);
  if (!PROVINCE_CODES.has(province)) {
    return `Código de provincia inválido (${province}). Debe estar entre 01 y 31`;
  }

  const thirdDigit = parseInt(clean.charAt(2), 10);
  if (thirdDigit > 5) {
    return "El tercer dígito debe ser 0-5 para personas naturales (6-9 solo para jurídicas)";
  }

  const coefficients = [2, 1, 2, 1, 2, 1, 2, 1, 2];
  let sum = 0;

  for (let i = 0; i < 9; i++) {
    const product = parseInt(clean.charAt(i), 10) * coefficients[i]!;
    sum += product >= 10 ? product - 9 : product;
  }

  const verifier = parseInt(clean.charAt(9), 10);
  const expectedVerifier = (10 - (sum % 10)) % 10;

  if (verifier !== expectedVerifier) {
    return "Cédula no válida. El dígito verificador no coincide";
  }

  return null;
}
